doc/design/mount-config-design.md

```diff
@@ -0,0 +1,115 @@
+| | |
+
+TODO: The source of private key for accessing an encrypted filesystem is not handled in this document at this time.
+
+There is no source location information in this configuration as this may change from one machine to the other.
+The source location will be passed via a command line argument to the insecure host application described in the next location.
+
+## Host configuration
+
+On the non-TEE host we need to specify the source mount point and associate it with the TEE target mount point.
+This needs to be more dynamic because each machine this is run on may have a different configuration.
+It is even possible that the source location may change with every run if the source location is generated with each run.
+
+The following new command line argument will be added:
+
+```text
+--mount <source>:<target>
+```
+
+| Name | Value |
+| -- | -- |
+| source | This is the source location of the mount in the insecure host. |
+| target | This is the mount point within the TEE. There needs to be an associated target mount configuration specified at signing for this location otherwise the mount will fail. |
+| | |
+
+The target location is part of the TEE measurement which specifies that the target directory is being accessed from an insecure source.
+This command line argument provides the mapping between the two.
+
+## Current mount capabilities
+
+An application run in Mystikos can call the SYS_mount syscall via the mount() CRT API to mount either an ext2, hostfs, or ramfs filesystems.
+The API specifies the source path, target path, filesystem type, mount flags and a data parameter.
+All three filesystems do not support mount flags.
+Only ext2 supports the data parameter which it extracts the key from.
+
+Other than the specification of the rootfs during launch there is no automatic way of creating a mount point inside the TEE without a the application itself calling the APIs.
+
+## Design
+
+During the signing of the application through sign-sgx or package-sgx commands via the myst command line tool the TEE mount configuration from the json configuration file is added to the signed application image and cannot be changed.
+
+
+On a host machine or container the application will be run either as an application in a directory with signed binaries, or via the self contained package binary.
+The new mount commandline parameter will be added to whatever command was used before, with one mount parameter option per mount point.
+
+The myst command will then launch the Mystikos runtime into the TEE and will pass the mount command line parameters in.
+
+Once the Mystikos kernel is entered the kernel will use the TEE-side JSON configuration and the command line parameters to call a set of myst_syscall_mount() calls, one for each mount point that is specified on the command line whos target path matches that of the mount JSON configuration.
```

Comment:
With the debug mode, where `myst` command is used to load a rootfs inside an enclave, does the mount commandline parameters require a config.json file with mount targets info be used in the commandline? Current `myst` commandline support skipping config.json. Should the debug mode skip the TEE-side configuration check if no mount target info is provided in config.json or the default config.json is implied (no config.json provided in Commandline)? For debug mode, if no mount target info is provided through config.json, I think the implementation can assume mounting flag is r/w.

Reply:
This is in fact a good question.
I am getting increasingly worried about the divergence between debug and signed/packaged mode at runtime in terms of environment variables and more.
I think the behaviour should be consistent across both so the config.json needs to be passed as a parameter and the mount parameters will behave the same in all three modes.

Reply:
I think it's reasonable to provide `-mount` option to users in debug mode with or without a config.json file.